File: backends/kotlin/tests/src/test/kotlin/com/sce/ecmascript/EcmaScriptSemanticsTest.kt
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
//
// Every engine this backend offers for `datamodel="ecmascript"`, measured
// against what ECMA-262 says.
//
// The expectations are not ours. They live in
// `tests/ecmascript/ecma262_semantics.json`, one claim per case with the
// clause that backs it, and the C++ engine test and the Rust frontend test
// read the same file. A per-backend copy would drift toward whichever engine
// reads it, which is the blindness this table was written to end: the same
// document answered differently depending on the engine, and no fixture in
// the repository could tell, because a W3C suite that is green end to end
// never asks `0 && x`.
//
// Why all three engines rather than the selected one. The C++ build picks an
// engine at configure time, so there the question is "did this build choose
// an ECMAScript engine". Kotlin has no such choice to make: a generated state
// machine takes its engine as a constructor argument, and this backend ships
// three — Rhino, QuickJS and Lua — each offered by `EngineFactory` as an
// equal option. So the question here is "which of the things we hand people
// are ECMAScript", and it can only be answered by asking all of them.
//
// The evaluation goes through `evaluateCondition` / `evaluateExpr`, which is
// what generated code calls (`scriptengine_helpers.kt.jinja2`), not a
// reimplementation of it.

package com.sce.ecmascript

import com.sce.runtime.ScxmlScriptEngine
import com.sce.scripting.RhinoScriptEngine
import com.sce.scripting.lua.LuaScriptEngine
import com.sce.scripting.quickjs.QuickJSScriptEngine
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import kotlin.test.Test
import kotlin.test.assertTrue

/**
 * Paths, relative to the repository root — the suite's `workingDir`, which is
 * also how the shared table is named below and how the C++ and Rust readers
 * name the same files.
 */
private const val DIVERGENCES_PATH = "tests/ecmascript/kotlin_lua_divergences.json"
private const val ENGINE_PATH =
    "backends/kotlin/lua/src/main/kotlin/com/sce/scripting/lua/LuaScriptEngine.kt"

/**
 * The `diverges_on` path this suite is the contract for.
 *
 * There are two routes from `datamodel="ecmascript"` into a Lua engine and they
 * fail differently: the engine's own input adapter rewriting the author's text,
 * which is what this suite exercises, and `sce-build`'s frontend having emitted
 * Lua at build time, which nothing on this backend does yet. Spelled the same
 * on all three readers of these lists — this one,
 * `tests/engine/LoweredEcma262Test.cpp` and
 * `sce-build/tests/ecma262_scoreboard_contract.rs`, which is where the
 * vocabulary is held against what the code generator derives.
 */
private const val RUNTIME_REWRITER_PATH = "runtime-rewriter"

/**
 * The text rewriter that used to be this engine's only answer, then its
 * fallback behind `sce-build`'s ECMAScript frontend, and is now retired.
 *
 * Two spellings because two questions are asked of them: whether the engine
 * still CALLS it (read from the class body, so a mention in a comment is not a
 * call) and whether the engine's documentation NAMES it.
 *
 * ⚠ They are kept after the retirement rather than deleted with it, because
 * the check they drive is what tells a rewriter that CAME BACK from one that
 * never left — and the arm they select is the one that would then have to be
 * satisfied. The tree-wide claim that no Kotlin file reaches the rewriter is
 * `retirement:kotlin-rewriter-deleted`'s, in
 * `sce-build/tests/lowering_decision_ledger.rs`; this pair is about the
 * engine's own documentation.
 */
private const val REWRITER_NAME = "EcmaScriptToLuaTransformer"
private const val REWRITER_CALL = "transformer."

/**
 * What stands where the fallback stood, and the clause that makes it an answer.
 *
 * `refusedToLower` is read from the class BODY and the clause from the header,
 * which is the same split the pair above uses and for the same reason: one asks
 * what the code does, the other what a reader is told.
 */
private const val REFUSAL_CALL = "refusedToLower("
private const val REFUSAL_CLAUSE = "§scxml-5.9.1"

/**
 * One case an engine answered differently from ECMA-262.
 *
 * Carried as a key plus a message rather than a message alone: the message is
 * for a person reading a failure, and the key is what a declared list can be
 * compared against. `source` alone is not the key — the table asks `a && b`
 * under two clauses, and collapsing them would let one declaration cover a
 * divergence nobody looked at.
 */
private data class Divergence(val source: String, val clause: String, val message: String) {
    val key: Key get() = Key(source, clause)
}

class EcmaScriptSemanticsTest {

    @Test
    fun rhinoAnswersWhatEcmaScriptAnswers() = measure("Rhino") { RhinoScriptEngine() }

    @Test
    fun quickJsAnswersWhatEcmaScriptAnswers() = measure("QuickJS") { QuickJSScriptEngine() }

    /**
     * Lua is measured too, and the assertion runs in both directions.
     *
     * "Both directions" is the part this test did not used to have. It
     * asserted only that the failure set was NOT EMPTY, which is satisfied by
     * one disagreement and by fifty, so the engine could regress or improve
     * for months without a word. Its own KDoc meanwhile carried "27 of its
     * 58" under a sentence saying this test held it to the measurement; it
     * held no number at all, and the shared table had since grown to 98 cases.
     * A declared list is what makes both directions visible — the same shape
     * `tests/ecmascript/lua_engine_divergences.json` gives the C++ selection,
     * and for the same reason its header states: a count that lives in prose
     * is a count nobody re-answers.
     *
     * ⚠ EMPTY IS A LEGAL ANSWER, and forbidding it was this test's own defect.
     * A third assertion here required the declared list to be non-empty, to
     * catch a list this suite had stopped reading — a real failure — but it
     * also failed on a list with nothing left in it, which is the terminal
     * state the whole seam is working towards. A counter whose zero is
     * forbidden is not a counter. What "was the list actually read" now means
     * is that the FILE was read, which [loadDeclaredDivergences] asserts, and
     * `ecma262_scoreboard_contract`'s `readers_of` is what still fails if
     * nothing opens the list at all.
     *
     * The two lists are separate on purpose, and that both are now empty is a
     * measurement rather than a definition: they were 23 and 44 while each
     * backend had its own text rewriter, and neither may be derived from the
     * other.
     */
    @Test
    fun theLuaEngineDivergesExactlyWhereItIsDeclaredTo() {
        val failures = collectFailures { LuaScriptEngine() }
        val declared = loadDeclaredDivergences()

        // Ordered before the floor deliberately. A list that is empty — the
        // first run of a new one — fails HERE, printing every divergence in
        // the shape the file takes, instead of dying below with nothing to
        // show the person who has to write it.
        val undeclared = failures.filterNot { it.key in declared }
        assertTrue(
            undeclared.isEmpty(),
            "${undeclared.size} expression(s) disagree with ECMA-262 on LuaScriptEngine " +
                "without being declared to. Either the engine lost an answer it used to " +
                "give, or $DIVERGENCES_PATH has not caught up with it. If it is the " +
                "second, these are the entries to add:\n" +
                undeclared.joinToString(",\n") {
                    "    { \"source\": ${jsonQuoted(it.source)}, \"clause\": ${jsonQuoted(it.clause)} }"
                } +
                "\n\nWhat each one answered:\n" +
                undeclared.joinToString("\n") { "  ${it.message}" },
        )

        val stillWrong = failures.map { it.key }.toSet()
        val repaired = declared.filterNot { it in stillWrong }
        assertTrue(
            repaired.isEmpty(),
            "${repaired.size} declared divergence(s) no longer describe this engine. Remove " +
                "them from $DIVERGENCES_PATH — a list that keeps a repaired case is a list " +
                "that cannot be trusted in the other direction either, and it is what a " +
                "reader consults to decide whether their document stays inside what this " +
                "engine covers.\n" +
                repaired.joinToString("\n") { "  ${it.source}  (${it.clause})" },
        )

        // Refutable, and refuted where a consumer can see it: what happens to
        // an expression the frontend REFUSES has to be readable from the
        // engine's own header. An empty divergence list says the frontend
        // answers the shared table's 98 cases; it says nothing about an
        // expression outside that table which the frontend refuses, and that is
        // the one a reader choosing this engine for `datamodel="ecmascript"`
        // has to know the fate of. A header claiming ECMAScript without saying
        // is the shape this file already paid for once, when it read "For
        // AOSP/AAOS production, this replaces Rhino with a faster native
        // engine" over a whole class of the table answered differently.
        //
        // ⚠ TWO STATES, BOTH WITH A REQUIREMENT — and that is the correction
        // this block carries. It used to be one-sided: while the fallback was
        // called the header had to name it, and it asked NOTHING otherwise, so
        // "derived, it retires itself" meant a retirement left the engine free
        // to document nothing about refusal at all. A check with an unasked arm
        // reads exactly like a live one. Now the retired state has its own
        // requirement: the call sites must reach `refusedToLower` and the
        // header must cite the clause that makes refusal the answer.
        val header = File(ENGINE_PATH)
        assertTrue(header.isFile, "the Lua engine is missing at ${header.absolutePath}")
        val engineSource = header.readText()
        val doc = engineSource.substringBefore("class LuaScriptEngine")
        val body = engineSource.substringAfter("class LuaScriptEngine")
        if (REWRITER_CALL in body) {
            assertTrue(
                REWRITER_NAME in doc,
                "$ENGINE_PATH still calls `$REWRITER_CALL` — the text rewriter is the " +
                    "fallback behind every lowering entry point — and its documentation does " +
                    "not name $REWRITER_NAME. An empty $DIVERGENCES_PATH means the frontend " +
                    "answers the ${loadSharedTable().size} cases of the shared table; it does " +
                    "not mean there is no second answer left. A reader choosing this engine " +
                    "for `datamodel=\"ecmascript\"` has to be able to see that an expression " +
                    "the frontend refuses is rewritten rather than refused.",
            )
        } else {
            assertTrue(
                REFUSAL_CALL in body,
                "$ENGINE_PATH no longer calls `$REWRITER_CALL`, so an expression the " +
                    "frontend refuses has to be REFUSED — and nothing in the class body " +
                    "calls `$REFUSAL_CALL`. Neither a rewrite nor a refusal means the text " +
                    "is going somewhere this test cannot see, which is the state a retiring " +
                    "round must not leave behind.",
            )
            assertTrue(
                REFUSAL_CLAUSE in doc,
                "$ENGINE_PATH refuses what the frontend refuses and its documentation does " +
                    "not cite $REFUSAL_CLAUSE. The clause is what makes the refusal an " +
                    "ANSWER rather than a defect — the processor places `error.execution` on " +
                    "the internal queue — and a reader choosing this engine has to be able " +
                    "to see that an expression outside what the frontend parses fails " +
                    "loudly rather than being guessed at.",
            )
        }

        // The header must point at the list rather than restate its size. The
        // count it used to carry outlived two growths of the shared table, and
        // a number in prose is the one thing here nothing can re-answer.
        assertTrue(
            DIVERGENCES_PATH in doc,
            "the Lua engine's documentation does not name $DIVERGENCES_PATH. That list is " +
                "the answer to the question its header raises — which cases — and a reader " +
                "who cannot reach it from the engine is left to take a count on trust, " +
                "which is how the previous one survived being wrong.",
        )
    }

    /**
     * The declared set, keyed the way the failures are.
     *
     * Read from disk rather than held in this file: the list is a measurement
     * a person maintains and a reader consults, and a Kotlin constant is
     * neither greppable beside the C++ list nor readable by anyone not
     * building this backend.
     */
    private fun loadDeclaredDivergences(): Set<Key> {
        val file = File(DIVERGENCES_PATH)
        assertTrue(
            file.isFile,
            "the declared divergences are missing at ${file.absolutePath}. This test compares " +
                "against them in both directions, so without the file it measures nothing.",
        )
        val root = Json.parseToJsonElement(file.readText()).jsonObject
        return root.getValue("divergences").jsonArray.mapNotNull { entry ->
            val obj = entry.jsonObject
            val key = Key(
                obj.getValue("source").jsonPrimitive.content,
                obj.getValue("clause").jsonPrimitive.content,
            )
            // Only the entries declared on THIS path. `diverges_on` splits the
            // list by which route into the Lua engine still answers the case
            // differently, and this suite hands the engine the author's
            // ECMAScript and has it lowered AT RUN TIME against the session's
            // scope — so it is the `runtime-rewriter` contract. The path keeps
            // that name because `EcmaScriptToLuaTransformer` is what used to do
            // that lowering; since it was retired the frontend does, and the
            // route is still the one that can be refused by a scope. The filter
            // is what keeps this from being an assumption: an entry that only
            // build-time lowering gets wrong would otherwise be reported here
            // as a repaired divergence and its deletion demanded.
            //
            // Unclassified is RED, not a default. An entry naming no path is
            // exempt from every per-path suite at once.
            val paths = obj["diverges_on"]
            assertTrue(
                paths != null,
                "the entry ${jsonQuoted(key.source)} / ${jsonQuoted(key.clause)} in $DIVERGENCES_PATH " +
                    "carries no `diverges_on`, so no per-path suite can tell whether it is about it.",
            )
            val onThisPath = paths!!.jsonArray.any { it.jsonPrimitive.content == RUNTIME_REWRITER_PATH }
            if (onThisPath) key else null
        }.toSet()
    }

    private fun measure(engineName: String, create: () -> ScxmlScriptEngine) {
        val failures = collectFailures(create)
        // Every case is reported, not just the first: a build that answers one
        // group wrong and another right is a different problem from one that
        // answers nothing, and the first failure alone cannot tell them apart.
        assertTrue(
            failures.isEmpty(),
            "${failures.size} of ${loadSharedTable().size} expressions disagree with ECMA-262, " +
                "evaluated by $engineName.\n" +
                "An engine offered for `datamodel=\"ecmascript\"` answers what that " +
                "language answers; one that does not is not a choice a consumer can " +
                "make safely, whatever else it is good at.\n" +
                failures.joinToString("\n") { it.message },
        )
    }

    private fun collectFailures(create: () -> ScxmlScriptEngine): List<Divergence> {
        // `loadSharedTable` carries the arity floor: a table that stopped
        // being read must not pass, and asserting it in one place keeps this
        // suite and `LoweredEcma262Test` on the same corpus by construction.
        val cases = loadSharedTable()

        val engine = create()
        val failures = mutableListOf<Divergence>()

        cases.forEachIndexed { index, case ->
            val sessionId = "ecma262_case_$index"
            engine.createSession(sessionId)
            try {
                engine.setupSystemVariables(sessionId, "ecma262")
                var setupOk = true
                if (case.setup.isNotEmpty()) {
                    try {
                        engine.executeScript(sessionId, case.setup)
                    } catch (failure: Exception) {
                        failures += Divergence(
                            case.source,
                            case.clause,
                            "[${case.source}] setup did not run: ${failure.message}" +
                                "\n  setup: ${case.setup}",
                        )
                        setupOk = false
                    }
                }
                if (setupOk) {
                    failures += evaluate(engine, sessionId, case)
                }
            } finally {
                engine.destroySession(sessionId)
            }
        }
        return failures
    }

    private fun evaluate(engine: ScxmlScriptEngine, sessionId: String, case: Ecma262Case): List<Divergence> {
        fun diverged(message: String) = listOf(Divergence(case.source, case.clause, message))

        if (case.asCondition) {
            val expected = case.expected
            val answered =
                try {
                    engine.evaluateCondition(sessionId, case.source)
                } catch (failure: Exception) {
                    return diverged(
                        "[${case.source}] failed to evaluate as a condition: " +
                            "${failure.message} (${case.clause})",
                    )
                }
            val wanted = (expected as? Answer.Bool)?.value
                ?: return diverged("[${case.source}] is a condition but names a non-boolean answer")
            return if (answered == wanted) {
                emptyList()
            } else {
                diverged(
                    "[${case.source}] answered $answered, ECMAScript says " +
                        "${expected.describe()} (${case.clause})",
                )
            }
        }

        val value =
            try {
                engine.evaluateExpr(sessionId, case.source)
            } catch (failure: Exception) {
                return diverged("[${case.source}] failed to evaluate: ${failure.message} (${case.clause})")
            }
        return if (answerMatches(value, case.expected)) {
            emptyList()
        } else {
            diverged(
                "[${case.source}] answered ${describeValue(value)}, ECMAScript says " +
                    "${case.expected.describe()} (${case.clause})",
            )
        }
    }

    /**
     * The row, the answer shapes, the number comparison and the shared
     * table's own floor live in `SharedEcma262Table.kt` beside this file:
     * `LoweredEcma262Test` asks the same table a different question — what
     * this backend's Lua answers for what the FRONTEND emitted — and a copy
     * of the reading in each suite would drift toward whichever one edits it,
     * exactly when a disagreement between the two is the interesting result.
     */
}

File: tests/engine/EcmaScriptSemanticsTest.cpp
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael

/**
 * @brief The engine this build selected answers what ECMA-262 says
 *
 * The C++ backend does not translate expressions. A generated state machine
 * carries the author's ECMAScript verbatim — `safeEvaluateGuard(engine,
 * session, "turns + 1 >= max_turns")` — and so does the Interpreter, so for
 * C++ the whole of `datamodel="ecmascript"` is whatever
 * `ScriptEngineProvider` hands back. That makes the engine, not the code
 * generator, the thing that has to be measured against the language.
 *
 * Which is why this reads `ScriptEngineProvider::getScriptEngine()` rather
 * than naming an engine: the assertion is "the engine this build ships is an
 * ECMAScript engine", and it has to be re-answered by every configuration of
 * `SCE_SCRIPT_ENGINE` rather than by the one a test author happened to pick.
 * `ECMAScriptComplianceTest` next door names `JSEngine` directly, so it
 * passes on a build that never runs `JSEngine`.
 *
 * The expectations are not ours. They live in
 * `tests/ecmascript/ecma262_semantics.json`, shared with
 * `sce-build/tests/ecmascript_semantics.rs`, and each one cites the ECMA-262
 * clause it comes from. A per-engine copy of a table drifts toward the engine
 * that reads it; one table cannot.
 */

#include "SCXMLTypes.h"
#include "common/GuardHelper.h"
#include "scripting/ScriptEngineProvider.h"
#include <cmath>
#include <fstream>
#include <gtest/gtest.h>
#include <memory>
#include <nlohmann/json.hpp>
#include <set>
#include <string>
#include <variant>
#include <vector>

#ifdef SCE_ENABLE_LUA
#include "scripting/LuaEngine.h"
#endif

namespace {

/// The shape `expect` takes in the shared table: exactly one of these keys.
struct Answer {
    enum class Kind { Bool, Number, String, Empty } kind;
    bool boolean = false;
    double number = 0.0;
    std::string text;

    std::string describe() const {
        switch (kind) {
        case Kind::Bool:
            return boolean ? "true" : "false";
        case Kind::Number:
            return std::to_string(number);
        case Kind::String:
            return "\"" + text + "\"";
        case Kind::Empty:
            return "null/undefined";
        }
        return "?";
    }
};

struct Case {
    std::string group;
    std::string setup;
    std::string source;
    bool asCondition = false;
    Answer expected;
    std::string clause;
};

Answer parseAnswer(const nlohmann::json &expect, const std::string &source) {
    Answer answer{};
    if (expect.contains("bool")) {
        answer.kind = Answer::Kind::Bool;
        answer.boolean = expect.at("bool").get<bool>();
    } else if (expect.contains("number")) {
        answer.kind = Answer::Kind::Number;
        answer.number = expect.at("number").get<double>();
    } else if (expect.contains("string")) {
        answer.kind = Answer::Kind::String;
        answer.text = expect.at("string").get<std::string>();
    } else if (expect.contains("empty")) {
        answer.kind = Answer::Kind::Empty;
    } else {
        // A case whose expectation cannot be read is not a case that passes.
        // Reading it as "no answer" would let a typo in a key name retire a
        // case silently, which is the failure mode the shared table exists to
        // remove.
        ADD_FAILURE() << "case [" << source << "] names no expected answer";
    }
    return answer;
}

std::vector<Case> loadCases() {
    std::ifstream file(SCE_ECMA262_CASES_PATH);
    if (!file.is_open()) {
        // Returning nothing rather than throwing keeps the verdict in the
        // test's own words: the floor below reports "0 cases", which says
        // the table was not read, where a stack trace would not.
        ADD_FAILURE() << "cannot read the shared table at " << SCE_ECMA262_CASES_PATH;
        return {};
    }

    nlohmann::json table;
    file >> table;

    std::vector<Case> cases;
    for (const auto &entry : table.at("cases")) {
        Case testCase;
        testCase.group = entry.value("group", std::string{});
        testCase.setup = entry.at("setup").get<std::string>();
        testCase.source = entry.at("source").get<std::string>();
        testCase.asCondition = entry.at("form").get<std::string>() == "condition";
        testCase.clause = entry.at("clause").get<std::string>();
        testCase.expected = parseAnswer(entry.at("expect"), testCase.source);
        cases.push_back(std::move(testCase));
    }
    return cases;
}

/// What the engine answered, in the same words the table uses.
std::string describe(const ::ScriptValue &value) {
    if (std::holds_alternative<::ScriptUndefined>(value)) {
        return "undefined";
    }
    if (std::holds_alternative<::ScriptNull>(value)) {
        return "null";
    }
    if (std::holds_alternative<bool>(value)) {
        return std::get<bool>(value) ? "true" : "false";
    }
    if (std::holds_alternative<int64_t>(value)) {
        return std::to_string(std::get<int64_t>(value));
    }
    if (std::holds_alternative<double>(value)) {
        return std::to_string(std::get<double>(value));
    }
    if (std::holds_alternative<std::string>(value)) {
        return "\"" + std::get<std::string>(value) + "\"";
    }
    if (std::holds_alternative<std::shared_ptr<::ScriptArray>>(value)) {
        return "[array]";
    }
    if (std::holds_alternative<std::shared_ptr<::ScriptObject>>(value)) {
        return "[object]";
    }
    return "[unknown]";
}

/// An engine may hold a whole number as an integer or as a double, and
/// ECMA-262 has one Number type — so both spellings answer a `number` case.
bool matches(const ::ScriptValue &actual, const Answer &expected) {
    switch (expected.kind) {
    case Answer::Kind::Bool:
        return std::holds_alternative<bool>(actual) && std::get<bool>(actual) == expected.boolean;
    case Answer::Kind::Number:
        if (std::holds_alternative<int64_t>(actual)) {
            return static_cast<double>(std::get<int64_t>(actual)) == expected.number;
        }
        if (std::holds_alternative<double>(actual)) {
            return std::abs(std::get<double>(actual) - expected.number) < 1e-9;
        }
        return false;
    case Answer::Kind::String:
        return std::holds_alternative<std::string>(actual) && std::get<std::string>(actual) == expected.text;
    case Answer::Kind::Empty:
        return std::holds_alternative<::ScriptUndefined>(actual) || std::holds_alternative<::ScriptNull>(actual);
    }
    return false;
}

/// One case that answered something other than what the language says.
///
/// Carries which case it was, not only the sentence about it: the second
/// engine's answers are compared against a declared list, and a list keyed on
/// a rendered message would have to be rewritten whenever the wording moved.
struct Disagreement {
    size_t caseIndex;
    std::string message;
};

/// Ask every case of one engine and collect what disagreed.
///
/// Every case is reported, not just the first: an engine that answers one
/// group wrong and another right is a different problem from one that answers
/// nothing, and the first failure alone cannot tell them apart.
///
/// The session id carries a prefix because two engines are asked in one
/// process and a session id is only unique within the engine that owns it.
std::vector<Disagreement> disagreements(SCE::IScriptEngine &engine, const std::string &prefix,
                                        const std::vector<Case> &cases) {
    std::vector<Disagreement> failures;

    for (size_t index = 0; index < cases.size(); ++index) {
        const Case &testCase = cases[index];
        const std::string sessionId = prefix + "_ecma262_case_" + std::to_string(index);
        if (!engine.createSession(sessionId, "")) {
            failures.push_back({index, "[" + testCase.source + "] no session"});
            continue;
        }

        bool setupOk = true;
        if (!testCase.setup.empty()) {
            auto setupResult = engine.executeScript(sessionId, testCase.setup).get();
            if (!setupResult.isSuccess()) {
                failures.push_back({index, "[" + testCase.source + "] setup did not run: " +
                                               setupResult.getErrorMessage() + "\n  setup: " + testCase.setup});
                setupOk = false;
            }
        }

        if (setupOk && testCase.asCondition) {
            // The production guard path, not a reimplementation of it: this is
            // what a `cond=` attribute reaches in both engines.
            auto answered = SCE::GuardHelper::evaluateGuard(engine, sessionId, testCase.source);
            if (!answered.has_value()) {
                failures.push_back(
                    {index, "[" + testCase.source + "] failed to evaluate as a condition (" + testCase.clause + ")"});
            } else if (*answered != testCase.expected.boolean) {
                failures.push_back({index, "[" + testCase.source + "] answered " + (*answered ? "true" : "false") +
                                               ", ECMAScript says " + testCase.expected.describe() + " (" +
                                               testCase.clause + ")"});
            }
        } else if (setupOk) {
            auto result = engine.evaluateExpression(sessionId, testCase.source).get();
            if (!result.isSuccess()) {
                failures.push_back({index, "[" + testCase.source + "] failed to evaluate: " + result.getErrorMessage() +
                                               " (" + testCase.clause + ")"});
            } else if (!matches(result.getInternalValue(), testCase.expected)) {
                failures.push_back({index, "[" + testCase.source + "] answered " + describe(result.getInternalValue()) +
                                               ", ECMAScript says " + testCase.expected.describe() + " (" +
                                               testCase.clause + ")"});
            }
        }

        engine.destroySession(sessionId);
    }

    return failures;
}

std::string joined(const std::vector<std::string> &failures) {
    std::string text;
    for (const auto &failure : failures) {
        text += failure + "\n";
    }
    return text;
}

std::vector<std::string> messagesOf(const std::vector<Disagreement> &failures) {
    std::vector<std::string> messages;
    messages.reserve(failures.size());
    for (const auto &failure : failures) {
        messages.push_back(failure.message);
    }
    return messages;
}

/// A case, named the way the declared list names one.
///
/// Source and clause together, because a source can appear twice: `!a` is
/// asked of `0` under 12.5.9 and of `null` under 7.1.2, and one of those two
/// answers correctly.
std::string identify(const Case &testCase) {
    return testCase.source + " | " + testCase.clause;
}

/// The cases the `lua` selection is declared to answer differently.
///
/// Read from a file rather than written here so that the sentence "this
/// engine is not an ECMAScript engine" is a list a reader can check against
/// the clauses it cites, and so that closing one is a deletion from data
/// rather than an edit to a test.
std::vector<std::string> loadDeclaredDivergences() {
    std::ifstream file(SCE_LUA_DIVERGENCES_PATH);
    if (!file.is_open()) {
        ADD_FAILURE() << "cannot read the declared divergences at " << SCE_LUA_DIVERGENCES_PATH;
        return {};
    }
    nlohmann::json document;
    file >> document;

    std::vector<std::string> declared;
    for (const auto &entry : document.at("divergences")) {
        declared.push_back(entry.at("source").get<std::string>() + " | " + entry.at("clause").get<std::string>());
    }
    return declared;
}

/// A floor, not an equality: adding a case must not have to touch this
/// number, but a table that stopped being read must not pass either.
void assertTableWasRead(const std::vector<Case> &cases) {
    ASSERT_GE(cases.size(), 55u) << "the shared ECMA-262 table produced only " << cases.size()
                                 << " case(s), so this is not measuring the corpus it claims to";
}

}  // namespace

TEST(EcmaScriptSemantics, TheSelectedEngineAnswersWhatEcmaScriptAnswers) {
    const auto cases = loadCases();
    assertTableWasRead(cases);

    auto &engine = SCE::ScriptEngineProvider::getScriptEngine();
    const auto failures = messagesOf(disagreements(engine, "provider", cases));

    EXPECT_TRUE(failures.empty()) << failures.size() << " of " << cases.size()
                                  << " expressions disagree with ECMA-262, evaluated by "
                                  << SCE::ScriptEngineProvider::getEngineName()
                                  << " (SCE_SCRIPT_ENGINE=" << SCE::ScriptEngineProvider::getEngineId()
                                  << ").\nIf this build did not choose that engine on purpose, its CMake"
                                     " cache predates the default: `SCE_SCRIPT_ENGINE` is a cache entry, so"
                                     " an existing build directory keeps whatever it was configured with."
                                     " Reconfigure it (-DSCE_SCRIPT_ENGINE=quickjs) or delete the cache.\n"
                                  << joined(failures);
}

#ifdef SCE_ENABLE_LUA
/// The other engine this backend ships answers the same table.
///
/// Named rather than provided, for the reason `DomReadSurfaceOnLuaEngine`
/// carries next door: `SCE_SCRIPT_ENGINE` is a compile-time choice and no
/// gate configures `lua`, so the engine reached by that selection would be
/// compiled by every build and measured by none — while remaining a listed,
/// validated value of the cache entry that a consumer may choose.
///
/// What it measures is the emission in `EcmaScriptToLuaTransformer`, not the
/// semantics: `ecma_semantics.lua` is a shared runtime asset that this engine
/// already loads and that the code generator's Lua backends already call, so
/// a disagreement here says the transformer wrote a bare Lua operator where
/// the shared definition of the ECMAScript one was in scope.
///
/// The verdict has two sides, because a one-sided one rots. Asking only
/// "nothing new disagrees" lets the declared list keep claiming a divergence
/// that has since been repaired, and the list is the only written account of
/// what this engine is not; asking only "the declared ones still disagree"
/// lets a new one arrive unremarked, which is how the count reached 44 while
/// a comment said 26.
TEST(EcmaScriptSemanticsOnLuaEngine, TheSecondEngineDivergesExactlyWhereItIsDeclaredTo) {
    const auto cases = loadCases();
    assertTableWasRead(cases);

    const auto declared = loadDeclaredDivergences();
    ASSERT_FALSE(declared.empty()) << "the declared divergences were not read from " << SCE_LUA_DIVERGENCES_PATH;

    auto &engine = SCE::LuaEngine::instance();
    const auto failures = disagreements(engine, "lua", cases);

    std::set<std::string> declaredSet(declared.begin(), declared.end());
    std::set<std::string> observed;
    std::vector<std::string> undeclared;
    for (const auto &failure : failures) {
        const std::string name = identify(cases[failure.caseIndex]);
        observed.insert(name);
        if (declaredSet.count(name) == 0) {
            undeclared.push_back(failure.message);
        }
    }

    EXPECT_TRUE(undeclared.empty())
        << undeclared.size() << " expression(s) disagree with ECMA-262 on LuaEngine without being declared to.\n"
        << "Either the transformer lost an answer it used to give, or a new case reached a rewriting pass that\n"
        << "cannot read it. If it is the second, `tests/ecmascript/lua_engine_divergences.json` is where it is\n"
        << "written down, with what closing it would take.\n"
        << joined(undeclared);

    // Names in the list that answer no case at all: a divergence declared
    // against a source or clause the shared table no longer carries would
    // otherwise sit there forever, describing nothing.
    std::set<std::string> named;
    for (const auto &testCase : cases) {
        named.insert(identify(testCase));
    }

    std::vector<std::string> stale;
    for (const auto &entry : declared) {
        if (named.count(entry) == 0) {
            stale.push_back(entry + "  (no such case in the shared table)");
        } else if (observed.count(entry) == 0) {
            stale.push_back(entry + "  (now agrees with ECMA-262)");
        }
    }

    EXPECT_TRUE(stale.empty()) << stale.size()
                               << " declared divergence(s) no longer describe this engine. Remove them from\n"
                                  "`tests/ecmascript/lua_engine_divergences.json`: a list that keeps a repaired\n"
                                  "entry understates the engine as surely as a missing entry overstates it.\n"
                               << joined(stale);
}
#endif
